packages/debugger-extension/src/index.ts

```diff
@@ -708,57 +708,60 @@ const sourceViewer: JupyterFrontEndPlugin<IDebugger.ISourceViewer> = {
     });
     const { model } = service;
 
-    const onCurrentFrameChanged = (
+    const onCurrentFrameChanged = async (
       _: IDebugger.Model.ICallstack,
       frame: IDebugger.IStackFrame
-    ): void => {
-      debuggerSources
-        .find({
-          focus: true,
-          kernel: service.session?.connection?.kernel?.name ?? '',
-          path: service.session?.connection?.path ?? '',
-          source: frame?.source?.path ?? ''
-        })
-        .forEach(editor => {
-          requestAnimationFrame(() => {
-            void editor.reveal().then(() => {
-              const edit = editor.get();
-              if (edit) {
-                Debugger.EditorHandler.showCurrentLine(edit, frame.line);
-              }
-            });
-          });
-        });
+    ): Promise<void> => {
+      if (!service.isStarted || !frame?.source?.path) {
+        return;
+      }
+      try {
+        const source = await service.getSource({ path: frame.source.path });
+        if (source) {
+          openSource(source, frame);
+        }
+      } catch (error) {
+        console.error('Failed to fetch source:', error);
+      }
     };
     model.callstack.currentFrameChanged.connect(onCurrentFrameChanged);
 
     const openSource = (
       source: IDebugger.Source,
-      breakpoint?: IDebugger.IBreakpoint
+      breakpointOrFrame?: IDebugger.IBreakpoint | IDebugger.IStackFrame
     ): void => {
       if (!source) {
         return;
       }
       const { content, mimeType, path } = source;
-      const results = debuggerSources.find({
-        focus: true,
-        kernel: service.session?.connection?.kernel?.name ?? '',
-        path: service.session?.connection?.path ?? '',
-        source: path
-      });
-      if (results.length > 0) {
-        if (breakpoint && typeof breakpoint.line !== 'undefined') {
+      if (breakpointOrFrame && typeof breakpointOrFrame.line !== 'undefined') {
+        const results = debuggerSources.find({
+          focus: true,
+          kernel: service.session?.connection?.kernel?.name ?? '',
+          path: service.session?.connection?.path ?? '',
+          source: breakpointOrFrame?.source?.path ?? ''
+        });
+
+        if (results.length > 0) {
           results.forEach(editor => {
             void editor.reveal().then(() => {
-              editor.get()?.revealPosition({
-                line: (breakpoint.line as number) - 1,
-                column: breakpoint.column || 0
-              });
+              const edit = editor.get();
+              if (edit) {
+                edit.revealPosition({
+                  line: (breakpointOrFrame.line as number) - 1,
+                  column: breakpointOrFrame.column || 0
+                });
+                Debugger.EditorHandler.showCurrentLine(
+                  edit,
+                  breakpointOrFrame.line as number
+                );
+              }
             });
           });
         }
         return;
```

Comment:
That `return` statement is not placed correctly, we need to `return` there **only** if we were able to open an editor. Otherwise we never open the document if it's not open already.

Reply:
Thanks @martinRenou It has been fixed